Internal Server Error, json doesn't get written to.

I wanted to try out this script but I ended up with an internal server error after a few runs, but the pages that did get processed were not written to the file.

```
python -m pockexport.export --secrets /Users/nyla/Documents/secrets.py >"/Users/nyla/Documents/Pocket-Export$(date -I).json"
/Users/nyla/pers_venv/lib/python3.10/site-packages/pockexport/exporthelpers/logging_helper.py:136: UserWarning: You might want to 'pip install colorlog' for nice colored logs
  warnings.warn("You might want to 'pip install colorlog' for nice colored logs")
[DEBUG   2024-10-11 00:54:57,253 __main__ export.py:37  ] retrieving from offset=0 (expected total=None)
[DEBUG   2024-10-11 00:54:57,962 __main__ export.py:37  ] retrieving from offset=30 (expected total=5000)
[DEBUG   2024-10-11 00:54:58,597 __main__ export.py:37  ] retrieving from offset=60 (expected total=5000)
[DEBUG   2024-10-11 00:54:59,169 __main__ export.py:37  ] retrieving from offset=90 (expected total=5000)
[DEBUG   2024-10-11 00:54:59,727 __main__ export.py:37  ] retrieving from offset=120 (expected total=5000)
[DEBUG   2024-10-11 00:55:00,390 __main__ export.py:37  ] retrieving from offset=150 (expected total=5000)
[DEBUG   2024-10-11 00:55:01,169 __main__ export.py:37  ] retrieving from offset=180 (expected total=5000)
[DEBUG   2024-10-11 00:55:01,832 __main__ export.py:37  ] retrieving from offset=210 (expected total=5000)
[DEBUG   2024-10-11 00:55:02,492 __main__ export.py:37  ] retrieving from offset=240 (expected total=5000)
[DEBUG   2024-10-11 00:55:03,576 __main__ export.py:37  ] retrieving from offset=270 (expected total=5000)
[DEBUG   2024-10-11 00:55:14,412 __main__ export.py:37  ] retrieving from offset=300 (expected total=5000)
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/Users/nyla/pers_venv/lib/python3.10/site-packages/pockexport/export.py", line 104, in <module>
    main()
  File "/Users/nyla/pers_venv/lib/python3.10/site-packages/pockexport/export.py", line 84, in main
    j = get_json(**params)
  File "/Users/nyla/pers_venv/lib/python3.10/site-packages/pockexport/export.py", line 74, in get_json
    return Exporter(**params).export_json()
  File "/Users/nyla/pers_venv/lib/python3.10/site-packages/pockexport/export.py", line 60, in export_json
    assert res.get('error') is None, res  # just in case
AssertionError: {'error': 'Internal server error'}
```

Do you know what went wrong?

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -3,7 +3,8 @@
 dynamic = ["version"]  # version is managed by setuptools_scm
 name = "pockexport"
 dependencies = [
-    "pocket",  # pocket API bindings
+    "pocket"  , # pocket API bindings
+    "tenacity", # retrying API requests
 ]
 requires-python = ">= 3.8"
 

--- a/src/pockexport/export.py
+++ b/src/pockexport/export.py
@@ -2,8 +2,16 @@
 from __future__ import annotations
 
 import json
+import logging
 
 import pocket  # type: ignore
+from tenacity import (
+    before_sleep_log,
+    retry,
+    retry_if_exception,
+    stop_after_attempt,
+    wait_exponential,
+)
 
 from .exporthelpers.export_helper import Json
 from .exporthelpers.logging_helper import make_logger
@@ -14,6 +22,11 @@ from .exporthelpers.logging_helper import make_logger
 ###
 
 logger = make_logger(__name__, level='debug')
+
+
+def _retry_if(e: BaseException) -> bool:
+    # seems to randomly occur at times?
+    return 'Internal server error' in str(e)
 
 
 class Exporter:
@@ -27,6 +40,19 @@ class Exporter:
         def get(self, **kwargs):
             pass
 
+        @retry(
+            retry=retry_if_exception(predicate=_retry_if),
+            wait=wait_exponential(max=60 * 10),  # 10 mins
+            stop=stop_after_attempt(10),
+            before_sleep=before_sleep_log(logger, logging.WARNING),
+        )
+        def get_with_retry(*args, **kwargs) -> Json:
+            res, _headers = get(*args, **kwargs)
+            error = res.get('error')
+            if error is None:
+                return res
+            raise RuntimeError(error)
+
         all_items: dict[str, Json] = {}
 
         first_res: Json | None = None
@@ -35,7 +61,7 @@ class Exporter:
         while True:
             offset = len(all_items)
             logger.debug(f'retrieving from {offset=} (expected {total=})')
-            res, _headers = get(
+            res = get_with_retry(
                 self.api,
                 images=1,
                 videos=1,
@@ -57,7 +83,6 @@ class Exporter:
             if first_res is None:
                 first_res = res
 
-            assert res.get('error') is None, res  # just in case
             total = int(res['total'])
 
             new_items: dict[str, Json] = res['list']

